test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMultipartMixedBatchOutputContextCongruenceTests.cs

```diff
@@ -0,0 +1,935 @@
+                }
+
+                await multipartMixedBatchWriter.WriteEndChangesetAsync();
+                await multipartMixedBatchWriter.WriteEndBatchAsync();
+            }
+
+            this.asyncStream.Position = 0;
+            var asyncResult = await new StreamReader(this.asyncStream).ReadToEndAsync();
+            asyncResult = Regex.Replace(asyncResult, batchGuidRegex, batchResponseBoundary);
+
+            var syncResult = await TaskUtils.GetTaskForSynchronousOperation(() =>
+            {
+                IODataResponseMessage syncResponseMessage = new InMemoryMessage { Stream = this.syncStream };
+                using (var messageWriter = new ODataMessageWriter(syncResponseMessage, this.writerSettings))
+                {
+                    var multipartMixedBatchWriter = messageWriter.CreateODataBatchWriter();
+                    multipartMixedBatchWriter.WriteStartBatch();
+                    multipartMixedBatchWriter.WriteStartChangeset("ec3a8d4f");
+
+                    var operationResponseMessage1 = multipartMixedBatchWriter.CreateOperationResponseMessage("1");
+                    operationResponseMessage1.StatusCode = 200;
+
+                    using (var messageWriter1 = new ODataMessageWriter(operationResponseMessage1, nestedWriterSettings))
+                    {
+                        var jsonLightWriter = messageWriter1.CreateODataResourceWriter(this.orderEntitySet, this.orderEntityType);
+                        jsonLightWriter.WriteStart(orderResource);
+                        jsonLightWriter.WriteEnd();
+                    }
+
+                    multipartMixedBatchWriter.WriteEndChangeset();
+                    multipartMixedBatchWriter.WriteStartChangeset("f46c46e2");
+
+                    var operationResponseMessage2 = multipartMixedBatchWriter.CreateOperationResponseMessage("2");
+                    operationResponseMessage2.StatusCode = 200;
+
+                    using (var messageWriter2 = new ODataMessageWriter(operationResponseMessage2, nestedWriterSettings))
+                    {
+                        var jsonLightWriter = messageWriter2.CreateODataResourceWriter(this.customerEntitySet, this.customerEntityType);
+                        jsonLightWriter.WriteStart(customerResource);
+                        jsonLightWriter.WriteEnd();
+                    }
+
+                    multipartMixedBatchWriter.WriteEndChangeset();
+                    multipartMixedBatchWriter.WriteEndBatch();
+                }
+
+                this.syncStream.Position = 0;
+                var result = new StreamReader(this.syncStream).ReadToEnd();
+
+                return Regex.Replace(result, batchGuidRegex, batchResponseBoundary);
+            });
+
+            Assert.Equal(asyncResult, syncResult);
+        }
+
+        #region Exception Cases
+
+        [Fact]
+        public async Task WriteMultipartMixedBatchRequestWithChangeset_APIsYieldSameResultForContentIdIsNull()

```

Comment:
Am confused,
What are we testing here that is different from this test `WriteMultipartMixedBatchRequest_APIsYieldSameResultForContentIdNull`

Reply:
Since the generated contentId is a random Guid, the value in the async result will be different from the value in the sync result so we use Regex to find and replace it with a know value before doing the comparison. It's the same thing we're doing with the batch boundary since it's also randomly generated

Reply:
The different is that we're not writing a changeset in `WriteMultipartMixedBatchRequest_APIsYieldSameResultForContentIdNull`
Which is why the methods are named:
- WriteMultipartMixedBatchRequest_APIsYieldSameResultForContentIdNull
- WriteMultipartMixedBatchRequestWithChangeset_APIsYieldSameResultForContentIdIsNull

Reply:
So with the changeset, we throw an exception?

Reply:
Yes. And the reason for that is, we don't write Content ID into the payload if there's no changeset which means that if it's not supplied we can let it slide. The `dependsOnIds` are only relevant when changesets are in play. Here's how the payload looks when there's no changeset vs when there's a changeset

**When there's no changeset - whether Content ID is supplied or not:**
```
--batch_732c7424-5d10-4b39-b46c-f850bcdc628d
Content-Type: application/http
Content-Transfer-Encoding: binary

POST http://tempuri.org/Orders HTTP/1.1
OData-Version: 4.0
Content-Type: application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8

{"Id":1,"Amount":13}
--batch_732c7424-5d10-4b39-b46c-f850bcdc628d
Content-Type: application/http
Content-Transfer-Encoding: binary

POST http://tempuri.org/Customers HTTP/1.1
OData-Version: 4.0
Content-Type: application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8

{"Id":1,"Name":"Customer 1"}
--batch_732c7424-5d10-4b39-b46c-f850bcdc628d--
```

**Where there's a changeset - Content ID is mandatory:**
```
--batch_e20b091a-23d7-4752-9b7c-ce55044cc36d
Content-Type: multipart/mixed; boundary=changeset_ec3a8d4f

--changeset_ec3a8d4f
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 1

POST http://tempuri.org/Orders HTTP/1.1
OData-Version: 4.0
Content-Type: application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8

{"Id":1,"Amount":13}
--changeset_ec3a8d4f
Content-Type: application/http
Content-Transfer-Encoding: binary
Content-ID: 2

POST http://tempuri.org/Customers HTTP/1.1
OData-Version: 4.0
Content-Type: application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8

{"Id":1,"Name":"Customer 1"}
--changeset_ec3a8d4f--
--batch_e20b091a-23d7-4752-9b7c-ce55044cc36d--
```

Reply:
Got it. Maybe you can add a comment in the tests to explain this

Reply:
Will do